packages/gatsby/src/commands/develop-process.ts

```diff
@@ -152,6 +152,16 @@ module.exports = async (program: IDevelopArgs): Promise<void> => {
   const app = express()
   const parentSpan = tracer.startSpan(`bootstrap`)
 
+  app.use((req, res, next) => {
+    try {
+      decodeURIComponent(req.path)
+    } catch (e) {
+      return res.status(500).send(`URI malformatted`)
```

Comment:
Not happy with the changes in `develop` yet.

Right now, we're catching `decodeURIComponent` errors, early in time(and responding with "URI malformed") but the ideal thing may be to load the dev 404 page. Is there a way to load that early in time in develop-process?

Reply:
Maybe instead of returning here we could try following (it might not work, just an idea)

```diff
-return res.status(500).send(`URI malformatted`)
+// req.path is not valid, so we assign it to dev 404 and allow future handlers to serve it
+req.path = `/dev-404-page/`
```

If it would show dev 404 and it actually works, we could use that, but if it wouldn't just work I would say we should just keep 500 as you have, only thing maybe adding more contextual information like:

```
.send(`"${req.path}" is not valid URL`)
```